New Control - Remote device picker

avknaidu opened the pull request:
Issue: #1182 
<!-- Link to relevant issue. All PRs should be asociated with an issue -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

Feature 

## What is the new behavior?

Remote Device Picker

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../readme.md#supported)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/Microsoft/UWPCommunityToolkit/blob/master/docs/.template.md). (for bug fixes / features)
- [x] Sample in sample app has been added / updated (for bug fixes / features)
    - [x] Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/Microsoft/UWPCommunityToolkit-design-assets)
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [x] Contains **NO** breaking changes


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+                m_remoteSystemWatcher.Start();
+            }
+
+            _progressRing.IsActive = true;
```
Need to null check this in case progress ring isn't part of a custom template.

michael-hawker commented on docs/controls/RemoteDevicePicker.md:
```diff
@@ -0,0 +1,72 @@
+
+## Requirements
+
+| Device family | Universal, 10.0.14393.0 or higher |
```
@nmetulev now that our min is 15063, should we be using that as the min in our docs or leaving it for the min bar if people want to recompile from source?

Code-ScottLe commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/KindToImageConverter.cs:
```diff
@@ -0,0 +1,59 @@
+            string finalvalue = string.Empty;
+            switch ((string)value)
+            {
+                case "Desktop":
```
RemoteSystem API return the kind of device via a string. We could use an enum, but it is a wasted effort and also a problem for maintaining because a new device category can be added at anytime.

https://docs.microsoft.com/en-us/uwp/api/windows.system.remotesystems.remotesystem.kind#Windows_System_RemoteSystems_RemoteSystem_Kind

michael-hawker replied:
@Code-ScottLe avknaidu added an enum to represent this, so I was just saying we could use it here instead to not have to maintain the names in two places.

Code-ScottLe replied:
That was *somewhere* on my note when I was testing last night... somewhere (left it home today). But you are right. I was thinking something like removing this enum all together and let the converter handle the unknown device kind with some default kind/string/image.

Code-ScottLe commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/KindToImageConverter.cs:
```diff
@@ -0,0 +1,59 @@
+    /// <summary>
+    /// Converter to convert Device Type to Icon
+    /// </summary>
+    public class KindToImageConverter : IValueConverter
```
Maybe RemoteSYstemKindToSymbolConverter. It is based on this

https://docs.microsoft.com/en-us/uwp/api/windows.system.remotesystems.remotesystem.kind#Windows_System_RemoteSystems_RemoteSystem_Kind

Code-ScottLe commented:
I had a few notes down last night on this, mostly it was good from the feature point of view, but a few tweaks here and there would be good. @michael-hawker points out some good stuffs on the UI/UX front too.

avknaidu commented:
@michael-hawker @Code-ScottLe Thanks for all the feedback. i have few hrs that i can spend tomorrow on this. Will update the PR.

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Gets List of All Remote Systems based on Selection Filter
+        /// </summary>
+        public ObservableCollection<RemoteSystem> RemoteSystems { get; private set; }
```
Why does this need to be exposed publicly and why as an ObservableCollection rather than an ICollection?

skendrot replied:
That doesn't change that this is exposed publicly and as an ObseravleCollection

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Gets or sets the Line Color on control Header. takes **SystemControlBackgroundAccentBrush** by default
+        /// </summary>
+        public Brush HeaderLineColor
```
Why does it even need to be settable? Should just be black/white or use the app theme, or allow the user to restyle

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Gets or sets the DeviceList Selection Mode. Defaults to ListViewSelectionMode.Single
+        /// </summary>
+        public ListViewSelectionMode DeviceListSelectionMode
```
Are there multiple lists? If not simplify name to `SelectionMode`

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Fired when the Remote Device Picker is Closed.
+        /// </summary>
+        public event EventHandler<RemoteDevicePickerEventArgs> RemoteDevicePickerClosed;
```
With previous statement to have an async method to return the result this should be removed along with the event args class

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+
+        private void RemoteDevicePicker_SecondaryButtonClick(ContentDialog sender, ContentDialogButtonClickEventArgs args)
+        {
+            ObservableCollection<RemoteSystem> selectedItems = new ObservableCollection<RemoteSystem>();
```
Why is an ObservableCollection needed here? a simple Collection or List should be fine

Code-ScottLe replied:
See above. This is needed as not all RemoteSystem instances will be returned all at once. They are returned via event fired within the RemoteSystemWatcher

skendrot replied:
No... This is the return value for being done

Code-ScottLe replied:
True. Was thinking about a different one.

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <inheritdoc />
+        protected override void OnApplyTemplate()
+        {
+            _listDevices = GetTemplateChild("PART_LISTDEVICES") as ListView;
```
Names of controls should be UpperCamelCase. Can remove the `PART_` prefix

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+            _progressRing = GetTemplateChild("PART_PROGRESS") as ProgressRing;
+
+            var enumval = Enum.GetValues(typeof(DeviceType)).Cast<DeviceType>();
+            _listDeviceTypes.ItemsSource = enumval.ToList();
```
How would this get localized?

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+    /// </summary>
+    public sealed class RemoteDevicePicker : ContentDialog
+    {
+        private Dictionary<string, RemoteSystem> DeviceMap { get; set; }
```
private member variable should be prefixed with an underscore and lowerCamelCase. `_deviceMap`

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        {
+            RemoteSystems = new ObservableCollection<RemoteSystem>();
+            DeviceMap = new Dictionary<string, RemoteSystem>();
+            _listDeviceTypes.SelectionChanged += ListDeviceTypes_SelectionChanged;
```
Should subscribe to the event when the object is set in the ApplyTemplate method

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+
+        private void RemoteDevicePicker_Loaded(object sender, RoutedEventArgs e)
+        {
+            RemoteSystems = new ObservableCollection<RemoteSystem>();
```
Set these two in the constructor or at declaration. No need for this event handler

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+            await Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
+            {
+                _progressRing.IsActive = true;
+                if (DeviceMap.ContainsKey(args.RemoteSystem.Id))
```
Why remove it just to add it afterward?

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        {
+            await Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
+            {
+                _progressRing.IsActive = true;
```
When does IsActive get set to false?

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+        {
+            await Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
+            {
+                _progressRing.IsActive = true;
```
When does IsActive get set to false?

skendrot commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.xaml:
```diff
@@ -0,0 +1,217 @@
+    
+    <converters:KindToImageConverter x:Key="KindToImageConverter" />
+
+    <Style x:Key="ComboBoxStyleDevicePicker" TargetType="ComboBox">
```
Why is a custom combobox style needed?

nmetulev replied:
I'm curious as well?

Code-ScottLe commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,210 @@
+
+        private async void RemoteSystemWatcher_RemoteSystemAdded(RemoteSystemWatcher sender, RemoteSystemAddedEventArgs args)
+        {
+            await Dispatcher.RunAsync(CoreDispatcherPriority.Normal, () =>
```
I thought we still have `DispatcherHelper.ExecuteOnUIThreadAsync` ? What are all of the others controls using these days for the toolkit?

Code-ScottLe commented:
@avknaidu I was about to type in a review from my notes, but with the amount of UX and code style changes, I think I would wait until you have a chance to finalize the UX because that could cause a change in the way you have to handle Rome-related code. Go ahead and review those UX changes requests and I'll jump in after :)

avknaidu commented:
@skendrot @michael-hawker I believe i addressed most of the review comments. There are few items that i did not complete in this push due to time constraints.

@Code-ScottLe Please add your review points so that i can look into those.

Code-ScottLe commented:
@avknaidu Thanks mate! But it will have to wait a bit, being flooded with works ATM :( . Give me a few days and I will definitely ping back!

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,190 @@
+    {
+        private Dictionary<string, RemoteSystem> DeviceMap { get; set; }
+
+        private ListView listDevices;
```
These 3 fields are already LowerCamelCase.

````
private ListView listDevices;
private ComboBox listDeviceTypes;
private ProgressRing progressRing;
````

Code-ScottLe replied:
if it is fields it needs to be `private ListView _listDevices;` , lowerCamelCase is for properties.

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,190 @@
+                RemoteSystemWatcher m_remoteSystemWatcher = RemoteSystem.CreateWatcher();
+                m_remoteSystemWatcher.RemoteSystemAdded += RemoteSystemWatcher_RemoteSystemAdded;
+                m_remoteSystemWatcher.RemoteSystemRemoved += RemoteSystemWatcher_RemoteSystemRemoved;
+                m_remoteSystemWatcher.RemoteSystemUpdated += RemoteSystemWatcher_RemoteSystemUpdated;
```
Yes. Instead of `EnumerationCompleted` Event, I am using `RemoteSystemsUpdated` Event to make the ProgressRing IsActive property to False. I explicitly chose to use this Event so that I can show the progress ring until the ListView is updated. Let me know if it is not a good idea.

Code-ScottLe replied:
RemoteSystemUpdated is going to be fired when there is an update to existing / already-discovered `RemoteSystem`, But my question here is how are you going to detect that the discovery has finished  without using `EnumerationCompleted` ? Also I haven't found any code in the control that stop the discovery, which we totally should do.

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,190 @@
+            listDeviceTypes = GetTemplateChild("PART_ListDeviceTypes") as ComboBox;
+            progressRing = GetTemplateChild("PART_ProgressRing") as ProgressRing;
+
+            var enumval = Enum.GetValues(typeof(DeviceType)).Cast<DeviceType>();
```
Change Made to use RemoteSystemKinds.

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,190 @@
+    /// </summary>
+    public sealed class RemoteDevicePicker : ContentDialog
+    {
+        private Dictionary<string, RemoteSystem> DeviceMap { get; set; }
```
Removed this unwanted dictionary. Code uses Linq now

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,190 @@
+        public async Task<List<RemoteSystem>> PickDeviceAsync()
+        {
+            List<RemoteSystem> remoteSystems = new List<RemoteSystem>();
+            await ShowAsync();
```
This change was part of initial review.

avknaidu commented:
@Code-ScottLe Do you think it is a good idea to add Spatial and Proximity discovery to the initialization? Maybe add'l properties?

Code-ScottLe commented:
@avknaidu Sorry for the late response. I think we should expose as much choice as possible, but also give it some default parameter so people don't have to type in too many args , but it is there if they need it.

JustinXinLiu commented:
@avknaidu first good stuff. I just have a few things to point out after playing with the control:

1. I'd love to have double-click to select a device.
2. There should be some margins around the "Done" button. Also, is "Select" a better word here?
3. We should have the `ModelDisplayName` (e.g. Surface Book 2) listed below the display name.
4. We should have the `Status` listed below the display name. We might want to grey out and disable unavailable ones too.
5. Why a `ContentDialog`? I'd want to be able to light-dimiss it, now I have to click the button to close it.

avknaidu commented:
@JustinXinLiu 

>I'd love to have double-click to select a device.

Made changes to the control where 
`SelectionMode` is `Single`, - No `SecondaryButton` will be shown. Picker will close automatically, 
`SelectionMode` is not `Single` - `SecondaryButton` will be shown. Double click on Item will close picker automatically. 
`SelectionMode` is not `Single` - `SecondaryButton` will be shown. You can select Multiple and close the picker manually by Clicking on `SecondaryButton`.
> 2. There should be some margins around the "Done" button. Also, is "Select" a better word here?

Added Margins. You can set the Secondary Button Text now from Picker. If nothing is Set, Default will be **Select**

> 3. We should have the ModelDisplayName (e.g. Surface Book 2) listed below the display name.

From Visual/UI Perspective, it was looking Odd. So added the `ModelDisplayName` as Tooltip. Let me know if it is still not a good idea.

> 4. We should have the Status listed below the display name. We might want to grey out and disable unavailable ones too.

Since ListViewItems cannot be made **Unclickable** we can make another filter property which shows only Active ones on demand.

> 5. Why a ContentDialog? I'd want to be able to light-dimiss it, now I have to click the button to close it.

You can press `Escape` on keyboard and control will be closed.

avknaidu commented:
@Code-ScottLe @nmetulev @skendrot @JustinXinLiu Or anyone who has some feedback about below.

On the name though. Is **RomeHelper** OK? or should i change it to **RemoteSystemsHelper** ?

Note that RomeHelper will have more filters ( SpatialProximity, KnownRemoteSystemCapabilities etc. ) 
RomeHelper will have its own page and documentaion.

avknaidu commented:
> 1. Multi-selection box should use CheckBox instead of Shift + select. The latter is not clear.

Made changes to show `CheckBox` if `SelectionMode` is `Multiple`.

> 2. Pressing Esc to exit the popup doesn't always work. I think we need a close button too.

You can easily add the close button by adding `PrimaryButtonText` during Initialization.

> 3. The tooltip only shows up when you click on the text, it should be on when you hover on the listview item. I would still create a grey text below the main one instead of tooltips, it's much clearer since I can see everything in one screen.

Added the `ModelDisplayName` Under the `DeviceName`

> 4. When you open up the popup, the order of the items changes a few times.

This is because the Enumeration of devices will take few seconds. So even though the list is populated, it will still enumerate so list will be updated.

> 5. By default, why should we allow it to select an inactive computer? The reason that I like disabling instead of hiding inactive ones is that this way I know exactly how many computers I have connected and which ones are unavailable so I can turn them on accordingly. You should be able to disable a ListViewItem by setting the IsEnabled property to False.

Made changes to Disable devices that are inactive.

> 6. It would be nice for the picker to display a message saying the connection is either established or failed, before closing up.

Picker will only show devices that are active and returns all devices that are selected to use with the Launcher. Correct me if i am wrong in understanding `the connection is either established or failed` .

nmetulev commented:
ping @JustinXinLiu

JustinXinLiu commented:
Hey @avknaidu could you please resolve the conflicts? After that I will take a look. Thanks!

avknaidu commented:
@JustinXinLiu Done

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/Package.appxmanifest:
```diff
@@ -1,6 +1,6 @@
-<Package xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10" xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest" xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10" IgnorableNamespaces="uap mp">
-  <Identity Name="52b9212c-97a9-4639-9426-3e1ea9c1569e" Publisher="CN=Nikola" Version="4.0.0.0" />
+<Package xmlns="http://schemas.microsoft.com/appx/manifest/foundation/windows10" xmlns:mp="http://schemas.microsoft.com/appx/2014/phone/manifest" xmlns:uap="http://schemas.microsoft.com/appx/manifest/uap/windows10" xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3" IgnorableNamespaces="uap mp uap3">
+  <Identity Name="52b9212c-97a9-4639-9426-3e1ea9c1569e" Publisher="CN=Nikola" Version="2.2.0.0" />
```
Why this change?

avknaidu replied:
This was a Merge conflict. Corrected now.

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/Package.appxmanifest:
```diff
@@ -27,22 +27,23 @@
         <uap:SplashScreen Image="Assets\UWPCommunityToolkitSampleAppSplashScreen.png" BackgroundColor="#3750D1" />
       </uap:VisualElements>
       <Extensions>
-        <uap:Extension Category="windows.protocol">
```
Same, this change seems unnecessary

avknaidu replied:
happened during merging conflicts. Moved to where it was before.

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/RemoteDevicePicker/RemoteDevicePickerCode.bind:
```diff
@@ -0,0 +1,8 @@
+    Title = "Pick Remote Device",
+    SecondaryButtonText = "Select",
+};
+remoteDevicePicker.SelectionMode = bool.Parse(SelectionCheckBox.IsChecked.ToString()) ? ListViewSelectionMode.Extended : ListViewSelectionMode.Single;
```
the bool check in the sample code is unnecessary, no need to complicate the example. Simple `remoteDevicePicker.SelectionMode = ListViewSelectionMode.Extended` would do

nmetulev commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/RemoteDevicePicker/RemoteDevicePickerControlPage.xaml.cs:
```diff
@@ -0,0 +1,35 @@
+        public RemoteDevicePickerControlPage()
+        {
+            InitializeComponent();
+            var selectionEnum = Enum.GetNames(typeof(ListViewSelectionMode)).Cast<string>().ToList();
```
Not sure if None or Extended makes sense for the selection mode. Should we have a new enum that only has Single and Multiple?

avknaidu replied:
Makes more sense.

JustinXinLiu replied:
Looks like esc to close works when you click in the content area. So maybe set a focus inside the popup?

avknaidu replied:
Noticed that. Will commit with focus.

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/Microsoft.Toolkit.Uwp.UI.Controls.csproj:
```diff
@@ -7,6 +7,7 @@
     <PackageTags>UWP Toolkit Windows Controls XAML Range WrapPanel Adaptive Markdown BladeView Blade CameraPreview Camera Carousel DockPanel DropShadow Expander GridSplitter HeaderedContent ImageEx InAppNotification InfiniteCanvas Master Details MasterDetails Menu Orbit Radial Gauge RadiaGauge RadialProgressBar Scroll ScrollHeader StaggeredPanel Staggered Tile UniformGrid Uniform Grid</PackageTag ...
   </PropertyGroup>
 
+
```
Remove extra space

avknaidu replied:
Done

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/Microsoft.Toolkit.Uwp.UI.Controls.csproj:
```diff
@@ -20,6 +21,15 @@
     <None Include="$(OutDir)\Design\$(MSBuildProjectName).Design.dll" Pack="true" PackagePath="lib\$(TargetFramework)\Design" />
   </ItemGroup>
+ 
+  <ItemGroup>
```
This is not necessary

avknaidu replied:
Done

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+        /// <summary>
+        /// Gets or sets List of All Remote Systems based on Selection Filter
+        /// </summary>
+        private ObservableCollection<RemoteSystem> RemoteSystems { get; set; }
```
Could this be a private field instead?

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+            nameof(HeaderLineBrush),
+            typeof(Brush),
+            typeof(RemoteDevicePicker),
+            new PropertyMetadata(Application.Current.Resources["SystemControlBackgroundAccentBrush"]));
```
The default should be moved to the template, don't use Application.Current.Resources

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+            deviceList.Add("All");
+            deviceList.Add("Unknown");
+
+            _listDeviceTypes.ItemsSource = deviceList.OrderBy(a => a.ToString());
```
Need to make sure you check for null - any element in the template could be removed by the developer.

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+            deviceList.Add("Unknown");
+
+            _listDeviceTypes.ItemsSource = deviceList.OrderBy(a => a.ToString());
+            _listDeviceTypes.SelectionChanged += ListDeviceTypes_SelectionChanged;
```
You need to first unsubscribe old event in case template is changed. For example, the patern is:

```
if (_element != null)
{
// unsubscribe events
}
_element = GetTemplateChild("name of element in template") as TypeOfElement;

if (_element != null)
{
  // set properties, subscribe to events
}

nmetulev commented on Microsoft.Toolkit.Uwp/Helpers/RomeHelper/RomeHelper.cs:
```diff
@@ -0,0 +1,80 @@
+        /// <summary>
+        /// Gets or sets List of All Remote Systems based on Selection Filter
+        /// </summary>
+        public ObservableCollection<RemoteSystem> RemoteSystems { get; set; }
```
the setter should be private

nmetulev commented on Microsoft.Toolkit.Uwp/Helpers/RomeHelper/RomeHelper.cs:
```diff
@@ -0,0 +1,80 @@
+    /// <summary>
+    /// Helper to List Remote Devices that are accessible
+    /// </summary>
+    public class RomeHelper
```
Should this be called RemoteDeviceHelper to match the name of the control?

avknaidu replied:
Changed it to `RemoteDeviceHelper`

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+            return devices;
+        }
+
+        internal IEnumerable<RemoteSystem> ReturnDevices()
```
Any reason why this is internal?

avknaidu replied:
made them private

nmetulev commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,191 @@
+
+        private void ListDeviceTypes_SelectionChanged(object sender, SelectionChangedEventArgs e) => UpdateList();
+
+        internal void UpdateList()
```
Why are these methods internal?

avknaidu replied:
made them private

avknaidu commented:
@nmetulev @skendrot @JustinXinLiu Ready for review.

avknaidu commented:
ping @nmetulev @skendrot @michael-hawker @JustinXinLiu @Code-ScottLe

nmetulev commented:
Sorry, most of us are busy this week on other projects, will review this as soon as possible.

avknaidu commented:
ping @nmetulev @skendrot @michael-hawker @JustinXinLiu @Code-ScottLe

skendrot commented:
Maybe github isn't doing a good job at showing reviewed items as changed, but most of my items above still show unresolved. I'll review again if they have been addressed

avknaidu commented:
I noticed that after you mentioned it. Noticed github put a word "Outdated" on them. 

![image](https://user-images.githubusercontent.com/6171200/47914130-77394b80-dec4-11e8-9902-d946b9d31893.png)

Code-ScottLe commented:
Last time I checked , you are good from Project-Rome functionalities wise. I quickly glance and look like nothing has changed since last time? If so, then I think you should be okay. I will let others walk through the design of the API first and I can jump in at the end to do the final verification. 

Sorry, it has been really hectic suddenly last 2 weeks, I even  missed my change to get my own feature in for the 5.0 release :(

nmetulev commented:
Just a quick review - looks much better now :)

* Canceling the picker while having devices selected still says devices were selected
* Alignment is broken on Advanced filter, looks weird
![image](https://user-images.githubusercontent.com/711864/48381457-c5063e80-e690-11e8-8d21-e9984b26e9cb.png)
* Select should be disabled when nothing is selected.
* On single selection, do not close popup on selection, still ask the user to click Select
* Is there a way to programmatically select the filter? For example, only show available devices
* When list is loading items move around and it's possible for items to shift while the user is clicking. Keep all the items in the position they show up

nmetulev commented on Microsoft.Toolkit.Uwp/Helpers/RemoteDeviceHelper/RemoteDeviceHelper.cs:
```diff
@@ -0,0 +1,164 @@
+        /// </summary>
+        public ObservableCollection<RemoteSystem> RemoteSystems { get; private set; }
+
+        private RemoteSystemWatcher remoteSystemWatcher;
```
add underscore for private fields

nmetulev commented on docs/controls/RemoteDevicePicker.md:
```diff
@@ -0,0 +1,55 @@
+
+## Requirements
+
+| Device family | Universal, 10.0.15063.0 or higher |
```
remove doc from here and create a new PR for docs here: https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs

nmetulev commented:
ping @avknaidu

michael-hawker commented:
@avknaidu thanks for the update, did you see our design feedback as well?

avknaidu commented:
@michael-hawker Yes sir. I considered all the design feedback and made changes. I will push the changes hopefully by this weekend. Working on some time sensitive stuff.

michael-hawker commented:
@avknaidu just wanted to check as I didn't see a comment from you, is this good to re-review?  Thanks! Hope you had a good holiday.

avknaidu commented:
@michael-hawker I wanted to complete the sample page for `RemoteDeviceHelper` and then comment that it is done, but was not able to do so in the last few days. However all the previous comments are addressed with last push. I will try to add the new page for Helper as soon as possible.

michael-hawker commented:
@avknaidu think you can get the sample page updated this week so this can make 5.1?

avknaidu commented:
@michael-hawker yes sir. Almost done. Preparing docs now and will push them.

avknaidu commented:
@michael-hawker @nmetulev @skendrot PR is ready for review.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,358 @@
+    /// <summary>
+    /// Picker Control to show List of Remote Devices that are accessible
+    /// </summary>
+    [TemplatePart(Name = "PART_ListDevices", Type = typeof(ListView))]
```
Looks like some template parts are missing that are used below?  Also, should pull out consts for the string names to use here and below like in the other controls.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,358 @@
+
+        private void RemoteDevicePicker_PrimaryButtonClick(ContentDialog sender, ContentDialogButtonClickEventArgs args)
+        {
+            _listDevices.SelectedItems.Clear();
```
Primary is cancel (maybe clarify names with comments or something)?  Need null check here (or mode check).  I'm getting a 'catastrophic failure' crash when I'm in single selection mode and hit Cancel, this is the line.

avknaidu replied:
Fixed this.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,358 @@
+            _listDevices = GetTemplateChild("PART_ListDevices") as ListView;
+            _listDeviceTypes = GetTemplateChild("PART_ListDeviceTypes") as ComboBox;
+            _progressRing = GetTemplateChild("PART_ProgressRing") as ProgressRing;
+            _commandSpace = GetTemplateChild("CommandSpace") as Grid;
```
I feel like you don't need this _commandSpace Grid reference just for a visibility.  If someone retemplated without a ListView, they can control whether they show buttons or not.

avknaidu replied:
Surrounded this with null check that way it will not break for retemplating.

avknaidu commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/samples.json:
```diff
@@ -304,6 +304,14 @@
+        "CodeFile": "RemoteDevicePickerCode.bind",
+        "Icon": "/SamplePages/RemoteDevicePicker/RemoteDevicePicker.png",
+        "DocumentationUrl": "https://raw.githubusercontent.com/Microsoft/UWPCommunityToolkit/master/docs/controls/RemoteDevicePicker.md"
+      },
```
Fixed.

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,358 @@
+
+        private void UpdateList()
+        {
+            var bindingList = new ObservableCollection<RemoteSystem>();
```
Rome API has `RemoteSystemWatcher_RemoteSystemAdded` and `RemoteSystemWatcher_RemoteSystemUpdated`. And both the events are fired when the watcher is initialized. That is the reason why i see the quick flicker in list load/reload.

michael-hawker replied:
Would it be better to have a copy of the remote system list, so when we get an update we can update the copy in place so that it doesn't disrupt the existing view?

avknaidu commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,358 @@
+        }
+
+        private void ListDevices_DoubleTapped(object sender, DoubleTappedRoutedEventArgs e)
+        {
```
You are absolutely correct. I propose to enable Double Tap only for Single Selection Mode. That way the confusion with multiple mode can be eliminated. and for single mode, it think the current way is correct because if you are double tapping on an item, you are try to select an item even though it is disabled and cannot be used.

avknaidu replied:
It was [this](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/2126#issuecomment-406105459) comment for which i have enabled double tap on the list. Better suggestion would be to remove the double tap completely to avoid confusion. ping @JustinXinLiu for his comment.

michael-hawker replied:
Yeah, Justin's offline at the moment, we can always add double-tap behavior later.  However, your latest change has the inverted logic (oops).

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,374 @@
+                _listDevices.IsMultiSelectCheckBoxEnabled = SelectionMode == RemoteDeviceSelectionMode.Multiple ? true : false;
+                if (_listDevices.SelectionMode != ListViewSelectionMode.Single)
+                {
+                    _listDevices.DoubleTapped += ListDevices_DoubleTapped;
```
You want this in the else so it double-taps in Single Mode, this enables it for Multiple currently.

avknaidu replied:
Fixed.

michael-hawker commented on Microsoft.Toolkit.Uwp.UI.Controls/RemoteDevicePicker/RemoteDevicePicker.cs:
```diff
@@ -0,0 +1,374 @@
+                    _listDevices.DoubleTapped += ListDevices_DoubleTapped;
+                    if (_commandSpace != null)
+                    {
+                        _commandSpace.Visibility = Visibility.Visible;
```
In single selection mode, I have no way to cancel when this is invisible now.

I'm not sure if `Esc` is supposed to work, but it may be an issue with focus?  I can't seem to keyboard focus on the dialog in the sample app to select an item with the keyboard?

avknaidu replied:
@michael-hawker I just tested this. `Esc works only if `CommandSpace` is enabled. That is the reason why in single mode it does not work. It works in multiple mode. I think the best way would be to show buttons by default and hide just press select and then access the system. Until we figure this out. I want to make sure stuff that is supposed to work a certain way does not break because of something that i am trying to make it work.

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/RemoteDevicePicker/RemoteDevicePickerControlPage.xaml.cs:
```diff
@@ -0,0 +1,35 @@
+                Title = "Pick Remote Device",
+                SelectionMode = (RemoteDeviceSelectionMode)Enum.Parse(typeof(RemoteDeviceSelectionMode), MyComboBox.SelectedValue.ToString()),
+                ShowAdvancedFilters = ShowAdvancedFilters.IsChecked.Value
+            };
```
Can we enable light-dismiss on this as well for the sample here?  Not sure if want check-box to show differences in behavior.

michael-hawker commented:
@avknaidu we're trying to close 5.1 today for new features, think you can get to the fix above today so we can ship this in 5.1?

avknaidu commented:
@michael-hawker  Yes. Working on it right now. Will push changes in next few mins. Question. Do you want me to remove DoubleTap for now? or keep it only for single mode?

michael-hawker commented:
@avknaidu I think we should support DoubleTap just for single selection mode.  But we should probably always have the Select/Cancel buttons visible, eh?  That provides in the standard case, a single selection and then confirmation (or they double-click), and it provides a clear way to cancel the single selection.  It also means we can remove the Command Grid as a templated part as we wouldn't need to toggle its visibility.

Sound good?

avknaidu commented:
@michael-hawker True. But in single selection mode, the dialog closes automatically as soon as you select an item. this was done upon request. Lets finalize the pieces here.

Mode: Single
Command Grid should be visible.
Do not close dialog as soon as you select an item and click Select to close the grid.
Double tap an item will close the dialog with that item selected.

Mode: Multiple
Command Grid will be visible.
Do not close dialog as soon as you select an item and click Select to close the grid.
Double tap an item will not close the dialog with that item selected.

By doing above, only thing that we will lose that we have right now is close on select a item. But i think you can use double tap anyways for that. Let me know if the above impl is ok and i will push the changes  within 20 mins of your reply.

I apologize but i want to do this right.

michael-hawker commented:
@avknaidu that sounds like a great plan!  Don't forget we'll be able to remove the command grid as a templated part with this too.

Sorry for the back and forth, we've been focused on accessibility this release, so I think the clarity in being able to see a way to cancel the single-select mode makes sense, and we'll have the double-click as a shortcut.

Appreciate all your efforts!  I'm going through your doc PR as well, so far just a couple of small suggestions there, I should be wrapped up with that in a couple of minutes.

avknaidu commented:
@michael-hawker expect changes in next few mins. Here and also for docs.

michael-hawker commented:
@avknaidu thanks a bunch! I'll test it out quick after I eat my lunch to be sure, but code looks good.

michael-hawker commented:
@avknaidu what OS are you running?  I'm seeing an issue with the RemoteDevicePicker sample on RS4 (17134). `"'input': Value must not be empty." Invalid JSON string exception.` when the content dialog is shown (the helper example works fine and shows the list of devices).

avknaidu commented:
@michael-hawker I'm on 17763,253.

michael-hawker commented:
@avknaidu I've tested Insider and 17763 as well.  I'm going to test RS3 (16299) in the morning and see if it's the same as well.  I have no idea where this message is coming from, just that it occurs internally to the XAML framework somewhere when the dialog is trying to show itself, so it must be some sort of templating thing, but I don't think I see OnApplyTemplate hit.

Assuming you don't have an easy way to test down-level, but if you think of anything let me know!

avknaidu commented:
I have a 16299 system but did not use it for past few months. I'll see if i can resurrect it and try to find the problem. will keep you posted.

michael-hawker commented:
@avknaidu tested on my 16299 VM as well, looks like they changed the API a bit and you're using a newer piece for `EnumerationCompleted`? I get this message for even the helper:

    "Unable to cast object of type 'Windows.System.RemoteSystems.RemoteSystemWatcher' to type 'Windows.System.RemoteSystems.IRemoteSystemWatcher2'."

If I comment line 90 with the EnumerationCompleted registration in RemoteDeviceHelper that example works.  I wonder if we need to do an API check for this event and then just use a timer otherwise.  I'm following up with the platform team on this.

The dialog also dies on 16299, but I haven't gotten closer to figuring that one out yet.  Will keep you posted if I find anything else about that one.

michael-hawker commented:
Also, make sure to do null checks before using all templated items, the `LoadFilters` method was missing a few. :)

michael-hawker commented:
If I comment out the `ScrollViewer` in the template (and fix all the null things), the dialog loads.  I can then uncomment it in the debugger and see the progress spinner.  So, need to dig in a bit more.  No idea what there wouldn't work on RS3/RS4 though.

michael-hawker commented:
Started a quick PR for null checks before I head back from the office for snow here: https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/2792

michael-hawker commented:
@avknaidu the culprit is the `SelectedIndex` property on ComboBoxes in the Style.  This isn't allowed until recently, there's some platform bug, which isn't obvious (see my related issue filed [here](https://github.com/dotnet/corefx/issues/35178)).

We'll have to set the index after we set the itemssource in code maybe or after each box has loaded?  It's going to be a pain but going to be the only way to support down-level.

Think you could investigate today?

I'm waiting to hear back on the suggested solution for `EnumerationCompleted` as well, that should be simpler to resolve though.

avknaidu commented:
@michael-hawker i was able to bring alive my old system and only removed the `SelectedIndex` from template for all 4 `ComboBox`es and added them to the code behind. And it works on 16299. I do not have 17134 to test though.  Can i push the changes to your branch directly so that it can be part of same PR?

michael-hawker commented:
@avknaidu sounds good, please make a push, I just have the one commit.  I can try testing on my VM again with all the changes.  Let me know if you need permissions to my branch or something.  Otherwise, feel free to branch off my branch and submit a new PR.

Did you see/fix the EnumerationCompleted issue I had seen as well?

avknaidu commented:
@michael-hawker. I did not see that error on 16299. Both Helper and Picker worked fine. Will push the commit in few mins.

michael-hawker commented:
@avknaidu sounds good, please make a push, I just have the one commit.  I can try testing on my VM again with all the changes.  Let me know if you need permissions to my branch or something.  Otherwise, feel free to branch off my branch and submit a new PR.

Did you see/fix the EnumerationCompleted issue I had seen as well?

[Ugh, GitHub didn't post this hours ago when I wrote it, but at least it saved it for me in my browser cache???]
